### PR TITLE
[FLINK-7757] [checkpointing] Introduce resource guard for RocksDBKeye…

### PR DESCRIPTION
**Pull request by StefanRRichter**
…dStateBackend to reduce locking and avoid blocking behavior.

## What is the purpose of the change

`RocksDBKeyedStateBackend` uses a lock to guard the db instance against disposal of the native resources while some parallel threads might still access db, which might otherwise lead to segfaults.

Unfortunately, this locking is a bit to strict and can lead to situations where snapshots block the pipeline. This can happen when a snapshot s1 is running and somewhere blocking in IO while holding the guarding lock. A second snapshot s2 can be triggered in parallel and requires to hold the lock in the synchronous part to get a snapshot from db. As s1 is still holding on to the lock, s2 can block here and stop the operator from processing further elements.
A simple solution could remove lock acquisition from the synchronous phase, because both, synchronous phase and disposing the backend are only allowed to be triggered from the thread that also drives element processing.

This PR removes long sections under the lock all together, to open up the possibility of parallel checkpointing. The change introduces a guard for the rocksdb instance that blocks disposal for as long as there are still clients potentially accessing the instance in parallel. This is realized by keeping a synchronized counter for active clients and block disposal until the client count drops to zero.

This approach could also be integrated with triggering timers, which have always been problematic in the disposal phase are currently unregulated. In the new model, they could register as yet another client.

## Brief change log

  - *Introduce a class `ResourceGuard` as client counter*
  - *Replaced use of `asyncSnapshotLock` with the guard*

## Verifying this change

This change is mostly covered by existing tests.

I added unit test `ResourceGuardTest` and `StateBackendTestBase::testParallelAsyncSnapshots` to ensure that parallel snapshots work.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @aljoscha

File: flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		synchronized (lock) {
+
+			closed = true;
+
+			while (clientCount > 0) {
+
+				try {
+					lock.wait();
+				} catch (InterruptedException e) {
+					Thread.interrupted();
```
Comment (GJL) [inline flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java]: This line caught my eye.
```
public static boolean interrupted()
```
> Tests whether the current thread has been interrupted. [...]

The boolean result is not used and when `InterruptedException` is thrown, the interrupt flag is cleared, so the method should return `false` here.

Did you intend to call `Thread.currentThread().interrupt()` and exit the `while` loop prematurely?

Comment (StefanRRichter) [reply flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java]: No, exactly this method was my intention. The API is very bad, but I am not interested in the result (because it is already clear here). What I am interested in is the side effect to clear the interrupted flag. If this is not cleared, future calls to `wait` trigger the interrupted exception again.

Comment (StefanRRichter) [reply flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java]: Just double checked, seems like I remembered incorrectly and this call is not needed here because the flag is reset already by the triggered exception.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java [hunk 2]
```diff
@@ -356,28 +354,26 @@ public int getKeyGroupPrefixBytes() {
 		final long checkpointTimestamp,
 		final CheckpointStreamFactory checkpointStreamFactory) throws Exception {
 
+		if (db == null) {
+			throw new IOException("RocksDB closed.");
```
Comment (bowenli86) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: This exception message looks a bit odd to me. Can you please enrich the message? E.g. "RocksDB has been closed expected. <suggestion for a solution>"?

**Comment by bowenli86** [comment]
The implementation of having s2 `acquire()` a lock while s1 still holds that lock seems problematic to me. Why not making it a `tryAcquire()`, as that in java semaphore?

**Comment by StefanRRichter** [comment]
@bowenli86 Can you go into detail what kind of problem you have in mind? This class does not work like a semaphore, i.e. `acquire` never truly blocks. It is just incrementing a count and does not represent a lock that can be held long time.

File: flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+	/**
+	 * Releases access for one client of the guarded resource. This method must only be called after a matching call to
+	 * {@link #acquireResource()}.
+	 */
+	public void releaseResource() {
+
+		synchronized (lock) {
+
+			if (STRICT_CHECKS && clientCount <= 0) {
```
Comment (bowenli86) [inline flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java]: What is the `STRICT_CHECKS` used for here? sorry, I haven't used invariant much in promgramming before

**Comment by bowenli86** [comment]
I was commenting on the scenario described in `What is the purpose of the change`. In that case, we want to either fail s2 or remove the lock for s2 as you did. Since we are required to support parallel snapshots, the latter should be the way to go.

LGTM 👍

File: flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java (hunk 3 above)
Comment (StefanRRichter) [reply flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java]: It is just a flag, whether or not to check that there cannot be more calls to `release()` than previous calls to `acquire`. Through the `static final boolean` in the condition, the jitter can eliminate the branch if it is deactivated.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java [hunk 4]
```diff
@@ -33,6 +32,7 @@
 import org.apache.flink.streaming.connectors.kafka.partitioner.FlinkKafkaPartitioner;
 import org.apache.flink.streaming.util.serialization.KeyedSerializationSchema;
 import org.apache.flink.util.NetUtils;
+import org.apache.flink.util.SerializableObject;
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java]: nit: unrelated change

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java [hunk 5]
```diff
@@ -618,19 +600,8 @@ public void releaseSnapshotResources() {
...
 			}
-		}
 
-		/**
-		 * Drop the created snapshot if we have ben cancelled.
-		 */
-		public void dropSnapshotResult() {
-			if (null != snapshotResultStateHandle) {
-				try {
-					snapshotResultStateHandle.discardState();
```
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: Cleanup is now handled somewhere else?

File: flink-contrib/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java [hunk 6]
```diff
@@ -215,46 +215,6 @@ public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
 	}
 
 	@Test
-	public void testRunningSnapshotAfterBackendClosed() throws Exception {
```
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java]: This test doesn't make sense anymore because a running snapshot operation will keep the backend form closing?

File: flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class is a guard for shared resources with the following invariants. The resource can be acquired by multiple
+ * clients in parallel through the {@link #acquireResource()} call. As a result of the call, each client gets a
+ * {@link Lease}. The {@link #close()} method of the lease releases the resources and reduces the client count in
+ * the {@link ResourceGuard} object.
+ * The protected resource should only be disposed once the corresponding resource guard is successfully closed, but
+ * the guard can only be closed once all clients that acquired a lease for the guarded resource have released it.
+ * Before this is happened, the call to {@link #close()} will block until the zero-open-leases condition is triggered.
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java]: nit "Before this has happened"

File: flink-core/src/test/java/org/apache/flink/util/ResourceGuardTest.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+					checker.set(false);
+				}
+			}
+		};
+
+		closerThread.start();
+
+		ResourceGuard.Lease lease_2 = resourceGuard.acquireResource();
+		lease_2.close();
+		Thread.sleep(50);
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/util/ResourceGuardTest.java]: I don't like sleep in tests but I don't see what else to do.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 9]
```diff
@@ -66,14 +66,13 @@
 import org.apache.flink.runtime.state.internal.InternalReducingState;
 import org.apache.flink.runtime.state.internal.InternalValueState;
 import org.apache.flink.runtime.util.BlockerCheckpointStreamFactory;
+import org.apache.flink.shaded.guava18.com.google.common.base.Joiner;
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: I don't think this will pass check style checks, if not, we still don't allow this usually. (I know that the import was already there previously 😅 )

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java (hunk 5 above)
Comment (StefanRRichter) [reply flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java]: Yes, the method was already unused before. Cleanup either happens already in the plain `close()` method of the checkpoint output stream, or in the `AsyncCheckpointRunnable::cleanup()` if ownership of a complete state handle was already transferred from the backend to the runnable.

File: flink-contrib/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java (hunk 6 above)
Comment (StefanRRichter) [reply flink-contrib/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java]: Exactly.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java (hunk 9 above)
Comment (StefanRRichter) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: IIRC, checkstyle actually _forced_ me to position the import there.

File: flink-core/src/test/java/org/apache/flink/util/ResourceGuardTest.java (hunk 8 above)
Comment (StefanRRichter) [reply flink-core/src/test/java/org/apache/flink/util/ResourceGuardTest.java]: I don't like them either. I will drop them because they are not actually adding real value and the test is clearer without them.

**Comment by StefanRRichter** [comment]
Thanks for the reviews @aljoscha @bowenli86 and @GJL . I addressed all points and will merge this now.
